### PR TITLE
i#1569 AArch64: Fix bug in encoding (SIMD structure load/store).

### PR DESCRIPTION
**Pull request by egrimley**
The "x16imm" operand type can be a register, or a constant denoted
in the encoding by register number 31. The value of the constant can
be determined from various bits in the instruction, but those bits
can not be uniquely determined from the constant operand. In particular,
we can not generate bit 30 in this case. Within the constraints of the
current formalism the best solution seems to be to remove bit 30 from
the bits generated by x16imm and duplicate the 14 bit patterns that
use x16imm.

There are other instances of such duplication. The formalism should
perhaps be improved.

Change-Id: I319d8e4b9550d39b11f5dce80fe7801115445446

### CODE REVIEW

**Comment by fhahn** [comment]
run aarch64 tests

File: core/arch/aarch64/codec.py [hunk 1]
```diff
@@ -135,7 +135,8 @@ def generate_encoder(patterns, opndsgen, opndtypes):
             vars = (['dst%d' % i for i in range(len(dsts))] +
                     ['src%d' % i for i in range(len(srcs))])
             c += ['    int opcode = instr->opcode;']
-            c += ['    uint ' + ', '.join(vars) + ';']
+            # The initial values are only required to silence a bad compiler warning:
+            c += ['    uint ' + ' = 0, '.join(vars) + ' = 0;']
```
Comment (fhahn) [inline core/arch/aarch64/codec.py]: I think it would be cleaner to add the initialisers in a separate commit.

Comment (egrimley) [inline core/arch/aarch64/codec.py]: You're probably right.

Comment (egrimley) [inline core/arch/aarch64/codec.py]: Separate pull request for initialisers: #2528

**Comment by egrimley** [comment]
run aarch64 tests

**Comment by fhahn** [comment]
Would it be hard to add a test case for that, e.g. encoding the newly added variants? I think this would be a helpful regression test if we decide to make changes to the formalism you hinted at.

Also the following sentence should be removed from the final commit message.

    This patch also adds some unnecessary initialisers to prevent certain
     compiler versions from breaking the build by incorrectly claiming that
     certain variables "may be used uninitialized".

**Comment by egrimley** [comment]
Yes, it ought to be possible to add a line to `dis-a64.txt`. I'll see if I can do that.

**Comment by egrimley** [comment]
I'll delete the last paragraph from the description. Presumably I'll have to delete it again when I squash and merge.

**Comment by egrimley** [comment]
rerun aarch64 tests

**Pull request merged**